optuna/integration/xgboost.py

```diff
@@ -9,6 +9,37 @@
     _available = False
 
 
+def _get_callback_context(env):
+    # type: (xgb.core.CallbackEnv) -> str
+    """return whether the current callback context is cv or train
+
+    Note:
+        `Reference
+        <https://github.com/dmlc/xgboost/blob/master/python-package/xgboost/callback.py>`_.
+    """
+
+    if env.model is None and env.cvfolds is not None:
+        context = 'cv'
+    else:
+        context = 'train'
+    return context
+
+
+def _remove_std_from_evaluation_result_list(xgb_callback_env):
+    # type: (xgb.core.CallbackEnv) -> None
+    """Custom XGBoost callback adapter to fix compatibility with Optuna pruner.
+
+    The Optuna pruner is expecting only two elements in each tuple of the evaluation_result_list.
+    It expects the observation_key and the evaluation metric only, but XGBoost is also providing
+    a third element: the stddev of the metric across the cross-valdation folds.
+    """
+
+    erl_orig = xgb_callback_env.evaluation_result_list
+    erl_no_std = [(key, metric) for key, metric, std in erl_orig]
+    erl_orig.clear()
+    erl_orig.extend(erl_no_std)
```

Comment:
Thank you for your advice. I will return new list in `_remove_std_from_evaluation_result_list` . But I have a question. How to wirte type hints? 

A data type of a second element of each tuples in `xgb_callback_env.evaluation_result_list` is numpy.float64 as default. How do I annotate this function.

```python
def _remove_std_from_evaluation_result_list(xgb_callback_env):
    # type: (xgb.core.CallbackEnv) -> List[Tuple[str, float]]
    # or
    # type: (xgb.core.CallbackEnv) -> List[Tuple[str, np.float64]]
    # or
    # type: (xgb.core.CallbackEnv) -> List[Tuple[str, Any]]
```

Reply:
> How to write type hints?

I prefer the first one in terms of simplicity.

```
# type: (xgb.core.CallbackEnv) -> List[Tuple[str, float]]
```

On the other hand, I think we can merge this function into `XGBoostPruningCallback.__call__` as follows because the actual logic will be only [L38](https://github.com/optuna/optuna/pull/865/files#diff-7fabbd1da1bd3bebde0e04965f81c892R38). What do you think?

```python
# Remove a third element: the stddev of the metric across the cross-valdation folds.
evaluation_result_list = [(key, metric) for key, metric, _ in env.evaluation_result_list]
```

Reply:
> On the other hand, I think we can merge this function into XGBoostPruningCallback.__call__ as follows because the actual logic will be only L38. What do you think?

You are absolutely right. I will merge this code into `XGBoostPruningCallback.__call__ `.